### PR TITLE
feat(telemetry): instrument NeMo-RL with OpenTelemetry via nemo-lens

### PR DESCRIPTION
**Pull request by ahmadki**
# What does this PR do ?

Adds **optional OpenTelemetry instrumentation** (distributed traces + metrics) across the NeMo-RL training loops, vLLM generation, and metrics logging, built on the shared **nemo-lens** library. Telemetry is fully optional — when nemo-lens is not installed (or `telemetry.enabled` is false) every instrumentation site degrades to a ~0-cost no-op via `nemo_rl/telemetry/_fallbacks.py`.

Highlights:
- **Telemetry module** (`nemo_rl/telemetry/`): `TelemetryConfig` (the `telemetry:` block of `MasterConfig`), the driver/worker lifecycle (`init_telemetry_driver` runs before `init_ray()` and exports `NEMO_RL_OTEL_*` env so Ray workers inherit it; raw env overrides YAML), `RLSpanGroup` presets, and the metrics tee.
- **Traces**: group-gated spans on GRPO (sync + async), PPO, SFT, DPO, RM and distillation — step / data_processing / rollout / generation / logprob / reward / advantage / policy(+value)_update / checkpoint / evaluate, plus a `job` span per trainer. Spans wrap the existing `timer.time()` blocks with no body re-indentation.
- **vLLM**: driver-side `rl.vllm.generate` / `generate_text` spans (nested under rollouts) + `gen_ai.*` token/latency metrics, plus opt-in vLLM-native OTLP engine tracing (`NEMO_RL_OTEL_VLLM_NATIVE_TRACING`).
- **Metrics**: `Logger.log_metrics` tees the standard `train`-prefix scalars (reward, loss, grad_norm, learning_rate, throughput, …) into `rl.*` metrics.
- **Deps / image**: new optional `telemetry` extra (`nemo-lens[sdk]`); the release Dockerfiles sync it into the driver venv so images ship telemetry-ready.
- **Tests + docs**: `tests/unit/telemetry/` and a 7-page `docs/observability/` guide (+ `nemo_rl/telemetry/README.md`).

Backend-agnostic: exports OTLP to any OTLP-compatible backend or an OpenTelemetry Collector via the standard `OTEL_EXPORTER_OTLP_*` env vars; console/JSON output via `NEMO_RL_OTEL_EXPORTER=console`.

# Issues

N/A 

# Usage

Enable telemetry and point it at a backend (or the console). No code changes to your run — just env or the `telemetry:` config block:

```bash
# Console (offline; prints spans + metrics as JSON to stdout)
NEMO_RL_OTEL_ENABLED=1 NEMO_RL_OTEL_EXPORTER=console NEMO_RL_OTEL_SPAN_GROUPS=per_step \
  uv run examples/run_grpo.py --config examples/configs/grpo_smoke.yaml

# Any OTLP backend / collector
export NEMO_RL_OTEL_ENABLED=1
export NEMO_RL_OTEL_SPAN_GROUPS=per_step
export OTEL_EXPORTER_OTLP_ENDPOINT=<your-otlp-endpoint>
uv run examples/run_grpo.py --config examples/configs/grpo_math_1B.yaml
```

Or in the run config:

```yaml
telemetry:
  enabled: true
  span_groups: per_step   # default | per_step | all
  exporter: otlp          # otlp | console
```

You then get `rl.grpo.*` spans (step, collect_rollouts, compute_logprobs, …), `rl.vllm.generate` spans, and `rl.*` metrics (`rl.reward.mean`, `rl.policy_loss`, `rl.grad_norm`, `rl.tokens_per_sec`, …).

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests? (`tests/unit/telemetry/` — span-group presets/resolution, lens-absent fallbacks, config→env translation, driver setup + resource attrs, metrics mapping, and in-memory span/metric assertions)
- [x] Did you run the unit tests and functional tests locally? (telemetry unit tests pass locally; functional/e2e still to run in CI)
- [x] Did you add or update any necessary documentation? (`docs/observability/` 7-page guide + `docs/index.md` toctree + `nemo_rl/telemetry/README.md`)

# Additional Information

- **Optional by design**: only `opentelemetry-api` is imported at module load; the SDK + exporters install via the `telemetry` extra / the release image. No hard runtime dependency, no import-time cost when disabled (gate check only).
- **Config**: `NEMO_RL_OTEL_*` env vars (with `NEMO_LENS_*` fallback) plus the `telemetry:` YAML block; raw env always overrides YAML. Endpoint / protocol / headers are the standard `OTEL_EXPORTER_OTLP_*` vars, so any backend works.
- **Companion change**: the `grad_norm` / `learning_rate` / `tokens_per_sec` metrics rely on additive gauges added to nemo-lens `record_rl_metrics` (separate nemo-lens change); without them those specific metrics are skipped and everything else is unaffected.
- **Scope notes**: DPO and RM only have the shared loop phases (no rollout/generation spans, since those loops have no such phase); dataset download/setup is not instrumented. vLLM native tracing is gRPC-only (needs a gRPC OTLP endpoint / collector), so it's opt-in and off by default.

### CODE REVIEW

File: nemo_rl/algorithms/grpo.py
Comment (rrs45) [inline nemo_rl/algorithms/grpo.py]: Is this missed?

File: nemo_rl/algorithms/grpo.py [hunk 1]
```diff
@@ -3894,12 +3979,27 @@ def async_grpo_train(
...
+                    tracer=_tracer,
+                    **{"rl.iteration": step + 1},
+                ),
+            ):
                 num_mask_sample_filtered = 0
 
                 # Sample trajectories from replay buffer
                 print("📦 Sampling from replay buffer...")
-                with timer.time("exposed_generation"):
+                with (
```
Comment (rrs45) [inline nemo_rl/algorithms/grpo.py]: This may not be the right place for this metric since for async, the rollout is occurring earlier in trajectory_collector,  here it is mainly collecting the buffer and not generation

File: nemo_rl/algorithms/grpo.py
Comment (rrs45) [inline nemo_rl/algorithms/grpo.py:4781]: Is this missed?

File: nemo_rl/algorithms/grpo.py (hunk 1 above)
Comment (ahmadki) [reply nemo_rl/algorithms/grpo.py]: Fixed

File: nemo_rl/models/generation/vllm/vllm_generation.py [hunk 2]
```diff
@@ -714,6 +756,7 @@ def generate_text(
                 f"Missing required keys for GenerationOutputSpec: {missing_keys}"
             )
 
+        _record_vllm_generation_metrics(self.cfg.get("model_name"), data, combined)
         return combined
```
Comment (rrs45) [inline nemo_rl/models/generation/vllm/vllm_generation.py:761]: Can we add support for async as well?

For ex: @trace_fn(RLSpanGroup.GENERATION, "rl.vllm.generate_async")

File: nemo_rl/models/generation/vllm/vllm_generation.py [hunk 3]
```diff
@@ -45,9 +45,48 @@
...
+
+        input_tokens = (
+            int(data["input_lengths"].sum()) if "input_lengths" in data else None
+        )
+        output_tokens = (
+            int(combined["generation_lengths"].sum())
+            if "generation_lengths" in combined
+            else None
+        )
+        record_inference_metrics(
```
Comment (rrs45) [inline nemo_rl/models/generation/vllm/vllm_generation.py:79]: can we add request duration request_duration_s?

File: nemo_rl/models/generation/vllm/vllm_worker.py [hunk 4]
```diff
@@ -77,6 +77,59 @@ def _context_capped_max_new_tokens(
...
+        return
+    if "otlp_traces_endpoint" not in supported:
+        logger.warning(
+            "nemo-lens: installed vLLM does not support 'otlp_traces_endpoint'; "
+            "skipping native vLLM tracing."
+        )
+        return
+    llm_kwargs.setdefault("otlp_traces_endpoint", endpoint)
+    if "collect_detailed_traces" in supported:
+        llm_kwargs.setdefault("collect_detailed_traces", ["all"])
```
Comment (rrs45) [inline nemo_rl/models/generation/vllm/vllm_worker.py:129]: Can we enable detailed traces with another env var like NEMO_RL_OTEL_VLLM_DETAILED_TRACES? As this could cause slowdowns

**Comment by ahmadki** [comment]
Closing in favor of https://github.com/NVIDIA-NeMo/RL/pull/3655
